modules/apps/object/object-service/src/main/java/com/liferay/object/internal/related/models/SystemObjectMtoMObjectRelatedModelsProviderImpl.java

```diff
@@ -0,0 +1,267 @@
+			Objects.equals(
+				objectRelationship.getDeletionType(),
+				ObjectRelationshipConstants.DELETION_TYPE_DISASSOCIATE)) {
+
+			_objectRelationshipLocalService.
+				deleteObjectRelationshipMappingTableValues(
+					objectRelationshipId, primaryKey);
+		}
+		else if (Objects.equals(
+					objectRelationship.getDeletionType(),
+					ObjectRelationshipConstants.DELETION_TYPE_PREVENT)) {
+
+			throw new RequiredObjectRelationshipException(
+				StringBundler.concat(
+					"Object relationship ",
+					objectRelationship.getObjectRelationshipId(),
+					" does not allow deletes"));
+		}
+	}
+
+	@Override
+	public void disassociateRelatedModels(
+			long userId, long objectRelationshipId, long primaryKey1,
+			long primaryKey2)
+		throws PortalException {
+
+		_objectRelationshipLocalService.
+			deleteObjectRelationshipMappingTableValues(
+				objectRelationshipId, primaryKey1, primaryKey2);
+	}
+
+	@Override
+	public String getClassName() {
+		return _systemObjectDefinitionMetadata.getModelClassName();
+	}
+
+	@Override
+	public String getObjectRelationshipType() {
+		return ObjectRelationshipConstants.TYPE_MANY_TO_MANY;
+	}
+
+	@Override
+	public List<T> getRelatedModels(
+			long groupId, long objectRelationshipId, long primaryKey, int start,
+			int end)
+		throws PortalException {
+
+		ObjectRelationship objectRelationship =
+			_objectRelationshipLocalService.getObjectRelationship(
+				objectRelationshipId);
+
+		PersistedModelLocalService persistedModelLocalService =
+			_persistedModelLocalServiceRegistry.getPersistedModelLocalService(
+				_systemObjectDefinitionMetadata.getModelClassName());
+
+		return persistedModelLocalService.dslQuery(
+			_getGroupByStep(
+				groupId, objectRelationship, primaryKey,
+				objectRelationship.isReverse(),
+				DSLQueryFactoryUtil.selectDistinct(_table)
+			).limit(
+				start, end
+			));
+	}
+
+	@Override
+	public int getRelatedModelsCount(
+			long groupId, long objectRelationshipId, long primaryKey)
+		throws PortalException {
+
+		ObjectRelationship objectRelationship =
+			_objectRelationshipLocalService.getObjectRelationship(
+				objectRelationshipId);
+
+		PersistedModelLocalService persistedModelLocalService =
+			_persistedModelLocalServiceRegistry.getPersistedModelLocalService(
+				_systemObjectDefinitionMetadata.getModelClassName());
+
+		return persistedModelLocalService.dslQueryCount(
+			_getGroupByStep(
+				groupId, objectRelationship, primaryKey,
+				objectRelationship.isReverse(),
+				DSLQueryFactoryUtil.countDistinct(
+					_table.getColumn(
+						_objectDefinition.getPKObjectFieldDBColumnName()))));
+	}
+
+	private GroupByStep _getGroupByStep(
+			long groupId, ObjectRelationship objectRelationship,
+			long primaryKey, boolean reverse, FromStep fromStep)
+		throws PortalException {
+
+		long objectDefinitionId1 = objectRelationship.getObjectDefinitionId1();
+		long objectDefinitionId2 = objectRelationship.getObjectDefinitionId2();
+
+		if (reverse) {
```

Comment:
@mateussgn we should take the reverse relationship here instead of changing ids

Reply:
Like the code we have here in this same PR:

```
ObjectRelationship objectRelationship =
			_objectRelationshipLocalService.getObjectRelationship(
				objectRelationshipId);

if (objectRelationship.isReverse()) {
	objectRelationship =
		_objectRelationshipLocalService.fetchReverseObjectRelationship(
			objectRelationship, false);
}
```

Reply:
does it make sense?

Reply:
@gabrielwas as objectRelationship argument is used inside a lambda function.
The lambda function does not accept variables that are not final or effectively final.
Show this message on an IntelliJ warning: `Variable used in lambda expression should be final or effectively final.`
And I can't deploy the module successfully.
As you recommended above, I'll try to get the relationship by its `relationshipId`.
This directs me to so another call on the database but will solve the actual approach.

Reply:
What do you think about it?